crates/red_knot/src/salsa_db.rs

```diff
@@ -0,0 +1,243 @@
+#![allow(unreachable_pub)]
+#![allow(clippy::used_underscore_binding)]
+
+use std::path::PathBuf;
+
+use salsa::{DebugWithDb, Event, EventKind, Storage};
+use tracing::{debug, debug_span, trace, warn};
+
+use crate::db::Upcast;
+use crate::salsa_db::source::File;
+
+use self::source::Db as SourceDb;
+
+pub mod lint;
+pub mod semantic;
+pub mod source;
+
+#[salsa::db(source::Jar, lint::Jar, semantic::Jar)]
+pub struct Database {
+    storage: Storage<Self>,
+
+    files: source::Files,
+}
+
+impl Database {
+    #[allow(unused)]
+    pub fn new() -> Self {
+        Self {
+            files: source::Files::default(),
+            storage: Storage::default(),
+        }
+    }
+}
+
+impl SourceDb for Database {
+    #[tracing::instrument(level = "debug", skip(self))]
+    fn file(&self, path: PathBuf) -> File {
+        self.files.resolve(self, path)
+    }
+}
+
+impl semantic::Db for Database {}
+
+impl Upcast<dyn source::Db> for Database {
+    fn upcast(&self) -> &(dyn source::Db + 'static) {
+        self
+    }
+}
+
+impl Upcast<dyn semantic::Db> for Database {
+    fn upcast(&self) -> &(dyn semantic::Db + 'static) {
+        self
+    }
+}
+
+impl lint::Db for Database {}
+
+impl salsa::Database for Database {
+    fn salsa_event(&self, event: Event) {
+        if matches!(event.kind, EventKind::WillCheckCancellation) {
```

Comment:
It's possible to create multiple snapshots of the database that then each can run in isolation (they still share the underlying caches). This is useful when using salsa in a multithreaded context. 

Now, Salsa cancels any pending snapshots (other threads) when you want to make changes to it. The way this works is that each query tests if cancellation was requested and if so, it panics with a specific error. The `WillCheckCancellation` indicates that Salsa now tests cancellation. 

I removed the log because it is very noisy. I think I often saw 2-3 of these logs per query. Maybe something that can be optimized later to reduce it to just one. Removing it made the log a bit more dense and easier to read thorough